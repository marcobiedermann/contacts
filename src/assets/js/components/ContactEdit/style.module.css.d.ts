declare const styles: {
  readonly form: string;
  readonly form__field: string;
  readonly form__input: string;
};
export = styles;
