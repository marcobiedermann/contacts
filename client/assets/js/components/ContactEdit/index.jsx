import { withFormik, Form, Field } from 'formik';
import PropTypes from 'prop-types';
import React from 'react';
import { translate } from 'react-i18next';
import { withRouter } from 'react-router-dom';
import { compose } from 'redux';
import styles from './style.css';

const ContactEdit = ({ values, isSubmitting }) => {
  const onSubmit = () => {
    const { removeContact, history, id } = values;

    removeContact(id);
    history.push('/contacts');
  };

  return (
    <Form>
      <div className={styles.form__field}>
        <label
          className={styles.form__label}
          htmlFor="name.firstName"
        >
          Firstname
        </label>
        <Field
          className={styles.form__input}
          id="name.firstName"
          name="name.firstName"
          type="text"
        />
      </div>

      <div className={styles.form__field}>
        <label
          className={styles.form__label}
          htmlFor="name.lastName"
        >
          Lastname
        </label>
        <Field
          className={styles.form__input}
          id="name.lastName"
          name="name.lastName"
          type="text"
        />
      </div>

      <div className={styles.form__field}>
        <label
          className={styles.form__label}
          htmlFor="address.streetAddress"
        >
          Street
        </label>
        <Field
          className={styles.form__input}
          id="address.streetAddress"
          name="address.streetAddress"
          type="text"
        />
      </div>

      <div className={styles.form__field}>
        <label
          className={styles.form__label}
          htmlFor="address.zipCode"
        >
          Zip
        </label>
        <Field
          className={styles.form__input}
          id="address.zipCode"
          name="address.zipCode"
          type="text"
        />
      </div>

      <div className={styles.form__field}>
        <label
          className={styles.form__label}
          htmlFor="address.city"
        >
          City
        </label>
        <Field
          className={styles.form__input}
          id="address.city"
          name="address.city"
          type="text"
        />
      </div>

      <div className={styles.form__field}>
        <label
          className={styles.form__label}
          htmlFor="address.country"
        >
          Country
        </label>
        <Field
          className={styles.form__input}
          id="address.country"
          name="address.country"
          type="text"
        />
      </div>

      <div className={styles.form__field}>
        <label
          className={styles.form__label}
          htmlFor="email"
        >
          Email
        </label>
        <Field
          className={styles.form__input}
          id="email"
          name="email"
          type="email"
        />
      </div>

      <div className={styles.form__field}>
        <label
          className={styles.form__label}
          htmlFor="phone"
        >
          Phone
        </label>
        <Field
          className={styles.form__input}
          id="phone"
          name="phone"
          type="tel"
        />
      </div>

      <div className={styles.form__field}>
        <button type="button" onClick={onSubmit}>
          {values.t('delete')}
        </button>
      </div>

      <div className={styles.form__field}>
        <input
          disabled={isSubmitting}
          type="submit"
          value={values.t('save')}
        />
      </div>
    </Form>
  );
};

ContactEdit.propTypes = {
  isSubmitting: PropTypes.bool,
  values: PropTypes.shape(),
};

ContactEdit.defaultProps = {
  isSubmitting: false,
  values: null,
};

export default compose(
  translate('common'),
  withRouter,
  withFormik({
    mapPropsToValues: ({
      t, updateContact, removeContact, history, id, name, address, email, phone,
    }) => ({
      t,
      updateContact,
      removeContact,
      history,
      id,
      name,
      address,
      email,
      phone,
    }),
    handleSubmit({
      updateContact,
      history,
      id,
      name,
      address,
      email,
      phone,
    }) {
      const contact = {
        name,
        address,
        email,
        phone,
      };

      updateContact(id, contact);
      history.push('/contacts');
    },
  }),
)(ContactEdit);
