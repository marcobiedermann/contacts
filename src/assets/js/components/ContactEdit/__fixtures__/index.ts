import { ContactEditProps } from '..';

const props: ContactEditProps = {
  initialValues: {
    address: {
      city: '',
      country: '',
      streetAddress: '',
      zipCode: '',
    },
    email: '',
    name: {
      firstName: '',
      lastName: '',
    },
    phone: '',
  },
  onSubmit: () => ({}),
};

export default props;
