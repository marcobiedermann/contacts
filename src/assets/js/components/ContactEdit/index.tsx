import { ErrorMessage, Field, FieldArray, Form, Formik, FormikHelpers } from 'formik';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { AddressProps } from '../Address';
import { EmailProps } from '../Email';
import Error from '../Error';
import Label from '../Label';
import { PhoneProps } from '../Phone';
import styles from './style.module.css';

export interface Values {
  addresses: AddressProps[];
  emails: EmailProps[];
  name: {
    firstName: string;
    lastName: string;
  };
  phones: PhoneProps[];
}

export interface ContactEditProps {
  initialValues: Values;
  onSubmit: (values: Values, formikHelpers: FormikHelpers<Values>) => void;
  validationSchema: any | (() => any);
}

function ContactEdit(props: ContactEditProps): JSX.Element {
  const { initialValues, onSubmit, validationSchema } = props;
  const { t } = useTranslation();

  return (
    <Formik initialValues={initialValues} onSubmit={onSubmit} validationSchema={validationSchema}>
      {(formikProps) => {
        const { isSubmitting, values } = formikProps;

        return (
          <Form className={styles.form}>
            <fieldset>
              <legend>Name</legend>

              <div className={styles.form__field}>
                <Label htmlFor="name.firstName">Firstname</Label>
                <Field className={styles.form__input} id="name.firstName" name="name.firstName" type="text" />
                <ErrorMessage name="name.firstName" component={Error} />
              </div>

              <div className={styles.form__field}>
                <Label htmlFor="name.lastName">Lastname</Label>
                <Field className={styles.form__input} id="name.lastName" name="name.lastName" type="text" />
                <ErrorMessage name="name.lastName" component={Error} />
              </div>
            </fieldset>

            <FieldArray name="addresses">
              {(arrayHelpers) => {
                const { push, remove } = arrayHelpers;

                return (
                  <fieldset>
                    <legend>Addresses</legend>

                    {values.addresses.map((_address, index) => {
                      const id = `addresses.${index}`;

                      return (
                        <div key={id}>
                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.type`}>Type</Label>
                            <Field as="select" className={styles.form__input} id={`${id}.type`} name={`${id}.type`}>
                              <option value="home">Home</option>
                              <option value="work">Work</option>
                            </Field>
                            <ErrorMessage name={`${id}.type`} component={Error} />
                          </div>

                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.value.street`}>Street</Label>
                            <Field
                              className={styles.form__input}
                              id={`${id}.value.street`}
                              name={`${id}.value.street`}
                              type="text"
                            />
                            <ErrorMessage name={`${id}.value.street`} component={Error} />
                          </div>

                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.value.zip`}>Zip</Label>
                            <Field
                              className={styles.form__input}
                              id={`${id}.value.zip`}
                              name={`${id}.value.zip`}
                              type="text"
                            />
                            <ErrorMessage name={`${id}.value.zip`} component={Error} />
                          </div>

                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.value.city`}>City</Label>
                            <Field
                              className={styles.form__input}
                              id={`${id}.value.city`}
                              name={`${id}.value.city`}
                              type="text"
                            />
                            <ErrorMessage name={`${id}.value.city`} component={Error} />
                          </div>

                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.value.country`}>Country</Label>
                            <Field
                              className={styles.form__input}
                              id={`${id}.value.country`}
                              name={`${id}.value.country`}
                              type="text"
                            />
                            <ErrorMessage name={`${id}.value.country`} component={Error} />
                          </div>

                          <button type="button" onClick={() => remove(index)}>
                            -
                          </button>
                        </div>
                      );
                    })}

                    <button
                      type="button"
                      onClick={() =>
                        push({
                          type: '',
                          value: {
                            city: '',
                            country: '',
                            street: '',
                            zip: '',
                          },
                        })
                      }
                    >
                      +
                    </button>
                  </fieldset>
                );
              }}
            </FieldArray>

            <FieldArray name="emails">
              {(arrayHelpers) => {
                const { push, remove } = arrayHelpers;

                return (
                  <fieldset>
                    <legend>Emails</legend>

                    {values.emails.map((_email, index) => {
                      const id = `emails.${index}`;

                      return (
                        <div key={id}>
                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.type`}>Type</Label>
                            <Field as="select" className={styles.form__input} id={`${id}.type`} name={`${id}.type`}>
                              <option value="home">Home</option>
                              <option value="work">Work</option>
                            </Field>
                            <ErrorMessage name={`${id}.type`} component={Error} />
                          </div>

                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.value`}>Email</Label>
                            <Field
                              className={styles.form__input}
                              id={`${id}.value`}
                              name={`${id}.value`}
                              type="email"
                            />
                            <ErrorMessage name={`${id}.value`} component={Error} />
                          </div>

                          <button type="button" onClick={() => remove(index)}>
                            -
                          </button>
                        </div>
                      );
                    })}

                    <button
                      type="button"
                      onClick={() =>
                        push({
                          type: '',
                          value: '',
                        })
                      }
                    >
                      +
                    </button>
                  </fieldset>
                );
              }}
            </FieldArray>

            <FieldArray name="phones">
              {(arrayHelpers) => {
                const { push, remove } = arrayHelpers;

                return (
                  <fieldset>
                    <legend>Phones</legend>

                    {values.phones.map((_phone, index) => {
                      const id = `phones.${index}`;

                      return (
                        <div key={id}>
                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.type`}>Type</Label>
                            <Field as="select" className={styles.form__input} id={`${id}.type`} name={`${id}.type`}>
                              <option value="home">Home</option>
                              <option value="work">Work</option>
                            </Field>
                            <ErrorMessage name={`${id}.type`} component={Error} />
                          </div>

                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.value`}>Phone</Label>
                            <Field className={styles.form__input} id={`${id}.value`} name={`${id}.value`} type="tel" />
                            <ErrorMessage name={`${id}.value`} component={Error} />
                          </div>

                          <button type="button" onClick={() => remove(index)}>
                            -
                          </button>
                        </div>
                      );
                    })}

                    <button
                      type="button"
                      onClick={() =>
                        push({
                          type: '',
                          value: '',
                        })
                      }
                    >
                      +
                    </button>
                  </fieldset>
                );
              }}
            </FieldArray>

            <div className={styles.form__field}>
              <button type="button">{t('common:delete')}</button>
            </div>

            <div className={styles.form__field}>
              <button disabled={isSubmitting} type="submit">
                {t('common:save')}
              </button>
            </div>
          </Form>
        );
      }}
    </Formik>
  );
};

export default ContactEdit;
