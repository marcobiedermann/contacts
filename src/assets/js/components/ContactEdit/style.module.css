.form {
  display: grid;
  gap: 1em;
}

.form__field {
  display: grid;
  gap: 1em;
  grid-template-columns: 1fr 5fr;
  grid-template-rows: min-content;
}

.form__input {
  border: 1px solid #d9d9d9;
  padding: 0.25em 0.5em;
}
