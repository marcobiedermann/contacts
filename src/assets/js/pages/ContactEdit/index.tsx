import { FormikHelpers } from 'formik';
import React, { FC, useCallback } from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';
import { Link, RouteComponentProps } from 'react-router-dom';
import { useFirestore, useFirestoreDocDataOnce } from 'reactfire';
import ContactEdit, { Values } from '../../components/ContactEdit';
import Loader from '../../components/Loader';
import validationSchema from '../../utils/validation';

interface Params {
  contactId: string;
}

interface Address {
  type: string;
  value: {
    city: string;
    country: string;
    street: string;
    zip: string;
  };
}

interface Email {
  type: string;
  value: string;
}

interface Name {
  firstName: string;
  lastName: string;
}

interface Phone {
  type: string;
  value: string;
}

interface Document {
  addresses: Address[];
  emails: Email[];
  id: string;
  name: Name;
  phones: Phone[];
}

const ContactEditPage: FC<RouteComponentProps<Params>> = (props) => {
  const {
    match: {
      params: { contactId },
    },
  } = props;
  const firestore = useFirestore();
  const ref = firestore.doc(`contacts/${contactId}`);
  const { status, data } = useFirestoreDocDataOnce<Document>(ref, {
    idField: 'id',
  });
  const { t } = useTranslation();

  const initialValues: Values = data;

  const onSubmit = useCallback((values: Values, formikHelpers: FormikHelpers<Values>) => {
    const { setSubmitting } = formikHelpers;

    ref.update(values);

    setSubmitting(false);
  }, []);

  if (status === 'loading') {
    return <Loader />;
  }

  return (
    <>
      <Helmet>
        <title>{t('pages:edit')}</title>
      </Helmet>
      <header>
        <Link to={`/contacts/${contactId}`}>{t('common:cancel')}</Link>
      </header>
      <ContactEdit initialValues={initialValues} onSubmit={onSubmit} validationSchema={validationSchema} />
    </>
  );
};

export default ContactEditPage;
