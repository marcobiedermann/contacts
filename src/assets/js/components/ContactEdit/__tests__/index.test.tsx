import { render } from '@testing-library/react';
import React from 'react';
import ContactEdit from '..';
import props from '../__fixtures__';

describe('contact edit component', () => {
  it('should render correctly', () => {
    expect.assertions(1);

    const { container } = render(<ContactEdit {...props} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
