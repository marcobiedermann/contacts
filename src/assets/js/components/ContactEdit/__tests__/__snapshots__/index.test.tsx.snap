// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contact edit component should render correctly 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="form"
  >
    <div
      class="form__field"
    >
      <label
        class="label"
        for="name.firstName"
      >
        Firstname
      </label>
      <input
        class="form__input"
        id="name.firstName"
        name="name.firstName"
        type="text"
        value=""
      />
    </div>
    <div
      class="form__field"
    >
      <label
        class="label"
        for="name.lastName"
      >
        Lastname
      </label>
      <input
        class="form__input"
        id="name.lastName"
        name="name.lastName"
        type="text"
        value=""
      />
    </div>
    <div>
      <div>
        <div
          class="form__field"
        >
          <label
            class="label"
            for="addresses.0.type"
          >
            Type
          </label>
          <select
            class="form__input"
            id="addresses.0.type"
            name="addresses.0.type"
          >
            <option
              value="home"
            >
              Home
            </option>
            <option
              value="work"
            >
              Work
            </option>
          </select>
        </div>
        <div
          class="form__field"
        >
          <label
            class="label"
            for="addresses.0.value.street"
          >
            Street
          </label>
          <input
            class="form__input"
            id="addresses.0.value.street"
            name="addresses.0.value.street"
            type="text"
            value=""
          />
        </div>
        <div
          class="form__field"
        >
          <label
            class="label"
            for="addresses.0.value.zip"
          >
            Zip
          </label>
          <input
            class="form__input"
            id="addresses.0.value.zip"
            name="addresses.0.value.zip"
            type="text"
            value=""
          />
        </div>
        <div
          class="form__field"
        >
          <label
            class="label"
            for="addresses.0.value.city"
          >
            City
          </label>
          <input
            class="form__input"
            id="addresses.0.value.city"
            name="addresses.0.value.city"
            type="text"
            value=""
          />
        </div>
        <div
          class="form__field"
        >
          <label
            class="label"
            for="addresses.0.value.country"
          >
            Country
          </label>
          <input
            class="form__input"
            id="addresses.0.value.country"
            name="addresses.0.value.country"
            type="text"
            value=""
          />
        </div>
        <button
          type="button"
        >
          -
        </button>
      </div>
      <button
        type="button"
      >
        +
      </button>
    </div>
    <div>
      <div>
        <div
          class="form__field"
        >
          <label
            class="label"
            for="emails.0.type"
          >
            Type
          </label>
          <select
            class="form__input"
            id="emails.0.type"
            name="emails.0.type"
          >
            <option
              value="home"
            >
              Home
            </option>
            <option
              value="work"
            >
              Work
            </option>
          </select>
        </div>
        <div
          class="form__field"
        >
          <label
            class="label"
            for="emails.0.value"
          >
            Email
          </label>
          <input
            class="form__input"
            id="emails.0.value"
            name="emails.0.value"
            type="email"
            value=""
          />
        </div>
        <button
          type="button"
        >
          -
        </button>
      </div>
      <button
        type="button"
      >
        +
      </button>
    </div>
    <div>
      <div>
        <div
          class="form__field"
        >
          <label
            class="label"
            for="phones.0.type"
          >
            Type
          </label>
          <select
            class="form__input"
            id="phones.0.type"
            name="phones.0.type"
          >
            <option
              value="home"
            >
              Home
            </option>
            <option
              value="work"
            >
              Work
            </option>
          </select>
        </div>
        <div
          class="form__field"
        >
          <label
            class="label"
            for="phones.0.value"
          >
            Phone
          </label>
          <input
            class="form__input"
            id="phones.0.value"
            name="phones.0.value"
            type="tel"
            value=""
          />
        </div>
        <button
          type="button"
        >
          -
        </button>
      </div>
      <button
        type="button"
      >
        +
      </button>
    </div>
    <div
      class="form__field"
    >
      <button
        type="button"
      >
        common:delete
      </button>
    </div>
    <div
      class="form__field"
    >
      <button
        type="submit"
      >
        common:save
      </button>
    </div>
  </form>
</DocumentFragment>
`;
